src/TypeChecker/Inference.hs

```diff
@@ -0,0 +1,521 @@
+inferExpr env e =
+    do  (s, t) <- infer e env
+        return (apply s t)
+
+
+-- TODO: Observation: Either environment should be extended 
+--                    (e.g. Map (T.Text, type) -> Scheme) - type as in TypeVar or TypeName etc.)
+--                    - solves checking for type/constructor/var declarations
+--                      but leaves exhaustive matching not feasable (i think)
+--                  OR same as above, but Map (T.Text, type, numOfConstr) -> Scheme
+--                     with numOfConstr only being used/useful for type == TypeName
+--                     this may solve the problem with exhaustive matching 
+--                  OR another data structure for types and their constructors is needed (what then?)
+typecheckTypeDef :: TDef -> Environment -> Either String Environment
+typecheckTypeDef (TDef typeName params kind cases _ ) env = 
+  -- add typeName to env / check for previous declaration as above
+  constructorsToEnv cases typeName params env
+
+
+  -- TODO: check for previous constructor declarations
+constructorsToEnv :: [TypeCase] -> TypeName -> [TypeVar] -> Environment -> Either String Environment
+constructorsToEnv [] (TypeName typeName) _ _ = 
+  Left $ "No type constructors for type: " ++ (T.unpack typeName)
+constructorsToEnv [(TypeCase (ConstructorName name) types _ )] 
+                  typeName 
+                  params 
+                  env =
+  case params of
+    [] -> 
+      Right $ extendEnv env (TypeVar name, schemeNoParams)
+    xs -> 
+      Right $ extendEnv env (TypeVar name, schemeParam) 
+  where 
+    schemeNoParams = Scheme [] (foldr FunctionType 
+                                      (NonPrimType typeName) 
+                                      types)
+    schemeParam = Scheme params 
+                         (foldr FunctionType 
+                                (ParamType typeName (map TypeVariable params)) types)
+
+constructorsToEnv ((TypeCase (ConstructorName name) types _ ):tail) 
+                  typeName 
+                  params 
+                  env =
+  case params of
+    [] -> 
+      let env' = extendEnv env (TypeVar name, schemeNoParams)
+      in constructorsToEnv tail typeName params env'
+    xs -> 
+      let env' = extendEnv env (TypeVar name, schemeParam)                               
+      in constructorsToEnv tail typeName params env'
+  where 
+    schemeNoParams = Scheme [] (foldr FunctionType 
+                                      (NonPrimType typeName) 
+                                      types)
+    schemeParam = Scheme params 
+                         (foldr FunctionType 
+                                (ParamType typeName (map TypeVariable params)) types)
+      
+
+typecheckAlias :: TAlias -> Environment -> Either String Environment
+typecheckAlias (TAlias (TypeName name) params kind alType _ ) env =
+  -- TODO: check if params are in type / check type correctness?
+  --        or is it done in parsing? 
```

Comment:
It's not.

Also the params don't need to appear in the type.

For example:
```
type Foo a where
  case Foo
```
is perfectly valid. It's a type of kind `* -> *`.